### PR TITLE
Limit cases of sygus inference based on type

### LINKED ISSUE #3250: CVC4 --sygus-inference assertion violation III
For the following formula, 
```
(declare-fun a () Real)   
(assert 
    (and  
        (and 
            (exists ((?b Real)) (forall ((?c Real)) (exists ((?d Real)) 
            (or  (and  (and  (and (and (< (+ (+ (+ 0 (* 68.0 ?c)) 0) (* 33.0 a)) 0.0) (<= 0 2.0)) 
            (or (<= 0 (+  (*  (+ (* 55.0 ?d) 0) (* 49.0 ?b)) 0))))))))))
        )
    )
) 

(check-sat)
```

CVC4 with `--sygus-inference gives 
```
(error "Assertion failure.
bool CVC4::theory::quantifiers::SynthConjecture::doCheck(std::vector<CVC4::NodeTemplate<true> >&)
/home/windomin/solver_binaries/cache/CVC4/src/theory/quantifiers/sygus/synth_conjecture.cpp:603:

  squery.isConst() && squery.getConst<bool>()
")
```

OS: Ubuntu 18.04
Revision: f727de33

### LINKED ISSUE #3356: Illegal argument detected using  --sygus-inference
Hello,

For this formula:
```

(declare-sort S 1)
(define-sort SB () (S Bool))
(declare-fun A () (S Bool))
(declare-fun B () SB)
(assert (= A B))
(check-sat)
(exit)
```

CVC4 returns the following error using the option --sygus-inference:
```
~/user/cvc4/build-2019-10-04-472c5a5/cvc4 --sygus-inference --produce-models --lang=smt2 test.smt2
(error "Illegal argument detected
CVC4::Expr CVC4::Datatype::mkGroundTerm(CVC4::Type) const

  `*this' is a bad argument
datatype is not well-founded, cannot construct a ground term!")
```

Omitting that option CVC4 reports that the formula is SAT:

OS: Linux 4.15.0-58-generic #64-Ubuntu SMP
Revision: 472c5a592c78e4757b3201f9e20908a4c645921b assertions:on

### PR DESCRIPTION
**Pull request by ajreynol**
This makes `--sygus-inference` a no-op for inputs where there is a free function whose sort cannot be handled in a sygus grammar.

It also fixes an issue where skolem variables were not being treated as functions-to-synthesize.

Fixes #3250 and fixes #3356.

### CODE REVIEW

File: test/regress/regress1/quantifiers/issue3250-syg-inf-q.smt2 [hunk 1]
```diff
@@ -0,0 +1,14 @@
+(set-logic ALL)
```
Comment (4tXJ7f) [inline test/regress/regress1/quantifiers/issue3250-syg-inf-q.smt2]: Shouldn't there be a `; COMMAND-LINE: --sygus-inference` for this test case, too?

Comment (ajreynol) [reply test/regress/regress1/quantifiers/issue3250-syg-inf-q.smt2]: It doesn't work with sygus inference :)
I added it just since it works without.

Comment (ajreynol) [reply test/regress/regress1/quantifiers/issue3250-syg-inf-q.smt2]: I've commented on this in the issue.

Comment (4tXJ7f) [reply test/regress/regress1/quantifiers/issue3250-syg-inf-q.smt2]: Oh, I see :) Alright, let's leave it for now.

**Pull request merged**